File: agent-service/src/core/agent-framework/base-agent.ts
import { AgentContext } from "./context";

// ── Status machine ─────────────────────────────────────────────────────────────
//
//   idle ──→ initializing ──→ ready ──→ running ──→ ready   (loop back)
//                │                        │
//                └──→ error               └──→ error
//   Any state ──→ stopped

export type AgentStatus =
  | "idle"
  | "initializing"
  | "ready"
  | "running"
  | "error"
  | "stopped";

/**
 * Abstract base class for all agents in the platform.
 *
 * Subclasses must implement three lifecycle hooks:
 *  - `onInit(ctx)`    – one-time setup (load config, warm up connections, etc.)
 *  - `onExecute(ctx)` – the main work for a single execution cycle
 *  - `onCleanup()`    – release resources (close connections, flush buffers, etc.)
 *
 * The public surface (`init`, `execute`, `cleanup`) wraps these hooks with
 * status transitions and error handling so subclasses stay focused on logic.
 */
export abstract class BaseAgent {
  private _status: AgentStatus = "idle";

  constructor(
    /** Unique instance identifier (e.g. UUID from the agent pool). */
    public readonly id: string,
    /** Agent type label used by the registry (e.g. "coder", "reviewer"). */
    public readonly type: string,
  ) {}

  // ── Public API ──────────────────────────────────────────────────────────────

  getStatus(): AgentStatus {
    return this._status;
  }

  /**
   * Initialise the agent. Safe to call multiple times — subsequent calls while
   * already `ready` are no-ops; calls while `running` throw.
   */
  async init(ctx: AgentContext): Promise<void> {
    if (this._status === "ready") return;
    if (this._status === "running") {
      throw new Error(`Agent ${this.id} is already running; cannot re-init.`);
    }
    this._status = "initializing";
    try {
      await this.onInit(ctx);
      this._status = "ready";
    } catch (err) {
      this._status = "error";
      throw err;
    }
  }

  /**
   * Execute the agent's main logic for the given context.
   * The agent must be in `ready` status before calling this.
   */
  async execute(ctx: AgentContext): Promise<void> {
    if (this._status !== "ready") {
      throw new Error(
        `Agent ${this.id} must be in "ready" state to execute (current: ${this._status}).`,
      );
    }
    this._status = "running";
    try {
      await this.onExecute(ctx);
      this._status = "ready";
    } catch (err) {
      this._status = "error";
      throw err;
    }
  }

  /**
   * Release all resources held by this agent instance.
   * Safe to call from any status; transitions to `stopped`.
   */
  async cleanup(): Promise<void> {
    try {
      await this.onCleanup();
    } finally {
      this._status = "stopped";
    }
  }

  // ── Lifecycle hooks (implement in subclasses) ───────────────────────────────

  protected abstract onInit(ctx: AgentContext): Promise<void>;
  protected abstract onExecute(ctx: AgentContext): Promise<void>;
  protected abstract onCleanup(): Promise<void>;
}

File: agent-service/src/agent_service/core/agent_framework/base_agent.py
from __future__ import annotations

import logging
from abc import ABC, abstractmethod
from typing import Literal, Optional

from agent_service.core.agent_framework.context import AgentContext

logger = logging.getLogger(__name__)

AgentStatus = Literal["idle", "initializing", "ready", "running", "error", "stopped"]


class AgentResult:
    """Structured result produced by the last successful execute() call."""

    def __init__(
        self,
        success: bool,
        output: str,
        error: Optional[str] = None,
        artifacts: Optional[list] = None,
    ) -> None:
        self.success = success
        self.output = output
        self.error = error
        self.artifacts = artifacts or []


class BaseAgent(ABC):
    """
    Abstract base class for all agents.

    Subclasses must implement three lifecycle hooks:
      - on_init(ctx)    — one-time setup
      - on_execute(ctx) — main work for a single execution cycle
      - on_cleanup()    — release resources
    """

    def __init__(self, id: str, type: str) -> None:
        self.id = id
        self.type = type
        self._status: AgentStatus = "idle"
        self._result: Optional[AgentResult] = None

    def get_status(self) -> AgentStatus:
        return self._status

    def get_result(self) -> Optional[AgentResult]:
        return self._result

    async def init(self, ctx: AgentContext) -> None:
        if self._status == "ready":
            return
        if self._status == "running":
            raise RuntimeError(f"Agent {self.id} is already running; cannot re-init.")
        self._status = "initializing"
        try:
            await self.on_init(ctx)
            self._status = "ready"
        except Exception:
            self._status = "error"
            raise

    async def execute(self, ctx: AgentContext) -> None:
        if self._status != "ready":
            raise RuntimeError(
                f'Agent {self.id} must be in "ready" state to execute (current: {self._status}).'
            )
        self._status = "running"
        try:
            await self.on_execute(ctx)
            self._status = "ready"
        except Exception:
            self._status = "error"
            raise

    async def cleanup(self) -> None:
        try:
            await self.on_cleanup()
        finally:
            self._status = "stopped"

    @abstractmethod
    async def on_init(self, ctx: AgentContext) -> None:
        ...

    @abstractmethod
    async def on_execute(self, ctx: AgentContext) -> None:
        ...

    @abstractmethod
    async def on_cleanup(self) -> None:
        ...
